### ISSUE TITLE
How can I improve the accuracy of yolov8 on custom data set

### ISSUE THREAD

**Issue by Hereguo** [ultralytics/ultralytics#6650]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Thank you for developing Yolov8! I'm sorry to trouble you, I have recently been doing research on object detection and have encountered some problems. My data set is about road crack detection, which is divided into different crack categories including repair. Now I add the data set and replace yolov8n with yolov8m model,changing batchsize and epoch and that has achieved the effect of mAP50-95 value of 0.48 on my verification set. I now want to further improve its map value, but I have no idea except model and layer changes. And trying to change learning rate is also almost impossible for me because I am just a student, can you give me some advice in this regard.

**Comment by glenn-jocher** [comment]
@Hereguo thank you for reaching out and for your kind words about YOLOv8. I'm glad to hear you've already achieved some promising results with your road crack detection dataset. To further improve your model's mAP, consider the following strategies:

1. **Data Augmentation**: Enhanced data diversity through augmentation techniques such as flipping, scaling, or color variation can help the model generalize better.

2. **Fine-tuning Hyperparameters**: Along with the learning rate, adjust the weight decay, momentum, or augmentations to optimize training.

3. **Quality Datasets**: Ensure your dataset has high-quality, well-labeled images. Include a variety of conditions and angles.

4. **Model Complexity**: If computational resources allow, using a heavier model (YOLOv8 'l' or 'x' versions) could help improve accuracy.

5. **Ensemble Models**: Combine predictions from different models or different checkpoints of the same model to boost performance.

6. **Advanced Loss Functions**: Experiment with different loss functions that might help focus the learning process on the most challenging aspects of your data.

7. **Transfer Learning**: Initiate training with weights from a model pre-trained on a similar task.

8. **Analyze Errors**: Review the predictions to understand where the model struggles. Sometimes, specific data or model adjustments are needed.

9. **Post-Processing**: Explore different non-maximum suppression (NMS) thresholds or confidence score thresholds.

For further reference, our documentation at https://docs.ultralytics.com covers detailed guidelines on model training and optimization. Keep experimenting and stay curious; breakthroughs often come through persistent iteration. Good luck with your research! 🚀🔍

**Comment by Hereguo** [comment]
Thank you very much for your reply. I will give it a try.

**Comment by glenn-jocher** [comment]
@Hereguo you're welcome! I'm glad I could help. Remember, persistence is key, and don't hesitate to visit our documentation for detailed guidance on each step. If you have any more questions during your experimentation, feel free to reach out. Good luck with your road crack detection project, and happy detecting! 🤖🛣️

**Comment by preyashyadav** [comment]
@glenn-jocher are there any preprocessing techniques that can help in increasing the model accuracy?

**Comment by glenn-jocher** [comment]
@preyashyadav absolutely! Preprocessing plays a critical role in improving model accuracy. Here are a few quick suggestions:

- **Normalization**: Ensure your images are normalized. This means scaling pixel values to be between 0 and 1 which can be automatically handled if you're using our dataset. For custom datasets, you might need to adjust this manually.
    ```python
    image /= 255.0
    ```
- **Resizing**: Consistent image sizes fed into the model ensure better performance. YOLOv8 handles this internally, but ensure your `imgsz` parameter is optimally set for your data.
- **Data Augmentation**: Techniques like flipping, rotation, and changing brightness/contrast can introduce variability, making your model more robust.

Just a quick tip, 🤖 be wary of excessive preprocessing as it could introduce noise or unwanted variances. Happy coding and best of luck with your accuracy improvements! 🚀

**Comment by cabral0413** [comment]
Currently, I'm working on a project that involves utilizing YOLOv8 for a classification task using a custom dataset stored in Google Drive. Additionally, I aim to evaluate the model's performance by computing validation accuracy and test accuracy. Furthermore, I'm interested in understanding if YOLOv8 includes any mechanisms to address overfitting during training. Lastly, for object detection purposes, I need to extract objects from images based on bounding box coordinates provided by YOLOv8. Here are my specific inquiries regarding these aspects:

Validation and Test Accuracy Computation:
How can I compute the validation accuracy and test accuracy while employing YOLOv8 for a classification task with my custom dataset stored in Google Drive?
Addressing Overfitting:
In the context of YOLOv8, if the model begins to overfit during training, are there any built-in mechanisms to automatically halt or mitigate the overfitting?
Object Extraction Using Bounding Boxes:
When utilizing YOLOv8 for object detection, how can I extract objects from images based on the bounding box coordinates provided by the model?

**Comment by glenn-jocher** [comment]
Hello! 🌟 Great questions! Let's dive right in:

**Validation and Test Accuracy:**
For computing validation and test accuracy with YOLOv8 on a custom dataset, ensure your dataset is appropriately structured and referenced in your `data.yaml` file. YOLOv8 will automatically calculate these metrics during the validation phase if you specify valid paths for both validation and test datasets. Here's an example snippet for `data.yaml`:

```yaml
train: /path/to/train/images
val: /path/to/val/images
test: /path/to/test/images

nc: 2  # number of classes
names: ['class0', 'class1']
```

**Addressing Overfitting:**
YOLOv8 employs various strategies to mitigate overfitting, such as data augmentation, dropout, and batch normalization. While there's no explicit early stopping mechanism, monitoring your training logs for signs of overfitting and adjusting your training epochs or employing techniques like model checkpointing based on validation loss can be beneficial.

**Object Extraction Using Bounding Boxes:**
To extract objects using bounding box coordinates provided by YOLOv8, you can simply crop the image based on these coordinates. Here's a quick Python snippet for cropping detected objects:

```python
import cv2

# Assuming 'results' is the output from your model's prediction
for *xyxy, conf, cls in results.xyxy[0]:
    x1, y1, x2, y2 = map(int, xyxy)
    crop_img = original_img[y1:y2, x1:x2]
    # Optionally, save or display `crop_img`
    cv2.imwrite('object.jpg', crop_img)
```

Remember to adapt paths and indexing based on your exact use case. Hope this helps, and feel free to reach out if you have more questions! Happy detecting! 🚀

**Comment by Saravanapandian224** [comment]
Hi Sir,
I have some improve accuracy for yolov8,
I've working on CCTV detection and Summary .  I am not getting accurate results for gender classification, and the YOLO person detection is adding extra counts. How can I fix this?

**Comment by glenn-jocher** [comment]
For improved accuracy in gender classification and reducing false positives in person detection, consider refining your dataset annotations and experimenting with model hyperparameters. Additionally, ensure your dataset is balanced and representative of the scenarios you aim to detect.

**Comment by zoegrace23** [comment]
Improving accuracy on a custom dataset often comes down to small but meaningful tweaks. Some of the best[ yolov8 improvements](https://yolov8.org/yolov8-improvements/) I’ve seen involve better data augmentation, cleaner labels, and training longer with a smaller learning rate. Curious what others here have tried?

**Comment by glenn-jocher** [comment]
@zoegrace23 great points! Those are definitely key strategies that work well in practice. Beyond what you mentioned, I've seen good results with [image tiling for small objects](https://docs.ultralytics.com/guides/model-evaluation-insights/#image-tiling-for-small-objects), [transfer learning from pre-trained weights](https://docs.ultralytics.com/guides/model-training-tips/#use-pre-trained-weights), and careful attention to [cross-validation techniques](https://docs.ultralytics.com/guides/kfold-cross-validation/) to ensure robust performance evaluation. The [model training tips guide](https://docs.ultralytics.com/guides/model-training-tips/) covers many of these strategies in detail for anyone looking to dive deeper.

**Comment by laxmanpagad21-design** [comment]
hello  @glenn-jocher   i am currently working on the PPE kit detection , model will get the input through the CC camera live feed  . for this i am training YOLOv8m.pt .and i am using the custom dataset .
MY QUERIES..............
1. what should be the dataset size for a moderate model 
2. how many annotations for each class
3.Can i train the model to detect the white shirt , is it possible , if yes then how ...

**Comment by glenn-jocher** [comment]
@laxmanpagad21-design for CCTV PPE with `yolov8m.pt`, a practical starting point is at least 100+ images and a few hundred labeled instances per class, with more data for confusing classes and difficult lighting/angles; yes, you can train `white_shirt` as its own class if you label it consistently across day/night, distance, and occlusion cases and include hard negatives, and the [data collection and annotation guide](https://docs.ultralytics.com/guides/data-collection-and-annotation/) is a good reference; since this thread is stale, please open a new issue or post in [Discussions](https://github.com/orgs/ultralytics/discussions) with sample frames and your `data.yaml` for more targeted help.

**Comment by varaprasad1016** [comment]
Hi,

I have a custom dataset created and curated with a single class (pigs) in this case and it is a live stock counting via RSTP. I was able to attain ~98%, however there is a consistent missing of a few. I fine tuned multiple parameters and thresholds. Introduced botsort, bytetrack too. However there is an issue with identifying the pigs when they are clustered in a batch. This is causing the loss of count and reduced accuracy. I did use YoloV8 segmentation originally to train my model. However, at this stage I am considering re-training of the model. As much as I understand the clustered pigs might still cause issue after re-training; as it might be due to memory leaks which is a different case, I would like to know the best possible approach for this use case and any suggested models. I am using the [NVIDIA Jetson AGX Orin 64GB Development Kit](https://uk.rs-online.com/web/p/processor-development-tools/2539662) and have resources enough to run the current use case. Any suggestions please ?

**Comment by glenn-jocher** [comment]
For clustered livestock, the main bottleneck is usually missed detections in crowded/overlapping scenes rather than the tracker, so I’d retrain with many more tightly labeled crowded examples, use higher `imgsz`, and prefer a segmentation model such as `yolo26m-seg.pt` or `yolo26l-seg.pt` on your Orin; since this issue is stale, please open a new issue with a few sample frames, your `data.yaml`, current `imgsz`/`conf`/`iou` settings, and the output of `yolo checks` so we can give targeted recommendations.
